### PR TITLE
fix(core): deduplicate messages by uuid in readSession (#137 phase 3)

### PR DESCRIPTION
**Pull request by DrumRobot**
## Summary

Resolves Issue #137 Phase 3 (Backend Deduplication / Data Integrity).

JSONL session files with duplicated `uuid` records — caused by file synchronization conflicts, repeated history appends, or cross-session edits — were inflating the rendered DOM cost and token-usage metrics. `readSession` now deduplicates messages by `uuid`, keeping only the **last occurrence**. Messages without a `uuid` (summary `leafUuid`, file-history-snapshot `messageId`, custom-title, agent-name) pass through unchanged.

The pure helper `dedupeMessagesByUuid` is exported via `packages/core/src/session/index.ts` so Issue #137 Phase 2 (UI duplicate warning banner) can reuse the same predicate in a downstream PR.

## Changes

- `packages/core/src/session/crud.ts` — new `dedupeMessagesByUuid` pure helper + `readSession` deduplicates before returning
- `packages/core/src/session/index.ts` — export `dedupeMessagesByUuid` for downstream consumers (Phase 2)
- `packages/core/src/__tests__/crud.test.ts` — 5 new tests in a `readSession dedup (Issue #137 Phase 3)` describe block

## Scope

- Phase 1 (UI hotfix — render-time key uniqueness) already shipped in PR #183
- Phase 2 (UI warning banner) intentionally deferred to a follow-up PR — Phase 2 requires either an API change on `/api/session` or a separate integrity endpoint; keeping this PR focused on data-integrity backend dedup keeps the diff small
- Issue #137 remains open after this PR; only the Phase 3 checkbox is checked

## Refs

Refs #137

## Test Plan

- [x] `pnpm -F @claude-sessions/core test` — 492/492 passing locally (33 in `crud.test.ts`, 5 new in the dedup describe block)
- [x] `pnpm -F @claude-sessions/core build` — succeeds (ESM + DTS)
- [x] `pnpm -F @claude-sessions/web build` — succeeds (isomorphic-core check; no SSR/CSR regression)
- [x] CI: full pnpm install + test on hosted runner — verified via gh pr checks 184 (test ubuntu pass 2m41s, test windows pass 4m10s)
- [x] CI: web e2e (playwright) — verified via gh pr checks 184 (e2e pass 1m18s) — no UI surface changed by this PR, but covered by existing specs
- [x] Manual smoke: covered by 5 new core unit tests with constructed dup-uuid fixtures (crud.test.ts:566-743) — exercises the same readSession code path as the web `/api/session` endpoint; web bundle build pass confirms no SSR/CSR regression

## Summary by CodeRabbit

* **New Features**
  * Session messages are now automatically deduplicated by unique identifier when loading, retaining only the most recent version of duplicate entries while preserving all non-duplicate messages.

* **Tests**
  * Added comprehensive test suite validating message deduplication behavior in session handling.

### CODE REVIEW

**Review by DrumRobot — COMMENTED** [review]
## Internal Code Review — [requesting-code-review](https://skills.sh/obra/superpowers/requesting-code-review)

> Reviewer matrix: **CodeRabbit** (walkthrough, no actionable comments) + **Internal Code Review** (Copilot unavailable on the acting account/org — auto-fallback per consolidate/pr.md Step 2.4).

### Strengths

1. **Plan alignment is exact.** Matches Issue #137 Phase 3 literally — dedupe by `uuid` at `crud.ts:readSession`, keep the last appearing record. Helper exported for Phase 2 reuse.
2. **Reader/writer separation preserved.** All mutation paths in `crud.ts` (`updateMessageContent`, `deleteMessage`, `splitSession`, `compressSession`, `restoreMessage`, `updateSessionSummary`) read JSONL directly via `readJsonlFile`, NOT through `readSession`. Writers see the original file lines (round-trip safe); only readers see the deduplicated view.
3. **Pure helper discipline.** `dedupeMessagesByUuid` has no I/O, no Effect wrapping, no side effects. Generic constraint `<T extends { uuid?: string }>` accepts both `Message` and generic records.
4. **Test data hygiene.** Generic UUIDs, generic project name, English text — conforms to `.claude/rules/tdd.md`.
5. **Behavior-focused tests.** All 5 tests exercise the public Effect-wrapped `readSession`, not the helper in isolation.
6. **Informative JSDoc.** Explains the *why* (Syncthing conflicts, repeated history appends, cross-session edits) and the *what* (keep last).
7. **No isomorphic-core violation.** No new `process`/`window`/Node built-ins. Helper is safe in the browser bundle.

### Issues

#### Critical (Must Fix)

None.

#### Important (Should Fix)

**1. Reordering can break parent-before-child chronology for interleaved-duplicate sessions** — `packages/core/src/session/crud.ts:132`

When duplicates are interleaved with their own children, dedup keeps the last occurrence in *its original position*. The interleaved test at `crud.test.ts:665-700` already pins this down:

```text
Before: [user-1, dup(first), assistant-1(parentUuid='dup'), dup(last)]
After:  [user-1, assistant-1, dup]     // 'dup' now comes AFTER its child
```

Downstream impacts:

- `validateChain` is fine (only checks uuid-set membership, not sequence)
- Sequential consumers (tree rendering / `MessageList.svelte`) that assume parents-before-children may render the orphaned child at the top
- For Syncthing conflicts with *divergent content* at the same uuid (rare but possible), the displayed conversation now shows the assistant replying to a different message body than the one it actually replied to

This is **inherent to the "keep last" rule chosen by the issue spec**, not a bug introduced by this PR. Recommendation: add a JSDoc paragraph noting the trade-off, and let Phase 2 (UI warning banner) surface divergent-content detection to the user.

**2. Missing chain-validity assertion on the interleaved-duplicate test** — `packages/core/src/__tests__/crud.test.ts:700`

The test verifies the new order `[user-1, assistant-1, dup]` but does not assert anything about chain validity. Adding one line would document that reordering is explicitly tolerated (because `validateChain` doesn't check sequence) and would catch a future regression if `validateChain` ever started enforcing sequence:

```ts
const chainResult = validateChain(messages)
expect(chainResult.valid).toBe(true)
```

#### Minor (Nice to Have)

3. **Empty array edge case not exercised** — `crud.test.ts`. `dedupeMessagesByUuid([])` works correctly (no-op), but an explicit `expect(dedupeMessagesByUuid([])).toEqual([])` makes the contract explicit.
4. **All-duplicate edge case not exercised** — `crud.test.ts`. The maximum-compression case (every record has the same uuid → returns 1 element) is worth pinning down.
5. **Type parameter widening note** — `crud.ts:132`. `<T extends { uuid?: string }>` is wider than `Message` (where `uuid: string`). Add a one-line JSDoc note: "Generic constraint allows non-`Message` callers (e.g., Phase 2 UI banner)."
6. **`forEach` + `filter` two-pass** — `crud.ts:132-138`. Cosmetic; current code is more readable than a single `reduce`. Not worth changing.
7. **Barrel-export comment** — `session/index.ts`. Add a JSDoc note on the re-exported `readSession` so package barrel consumers know the read is filtered.

### Recommendations

1. **Phase 2 helper coordination**: Phase 2 will need the *original count* of duplicates to render the warning banner. Either export a sibling `countDuplicateUuids(messages: T[]): number` now or leave it for Phase 2. Phase 2 cannot derive it from a *server-deduplicated* array.
2. **Document the reader/writer asymmetry**: The behavior — `readSession` dedupes but writers see the full file — is correct but subtle. Worth a one-line comment near `readSession` definition.
3. **Backward compatibility**: Sessions with no duplicate uuids are passthrough (every uuid maps to itself, no records filtered). The CodeRabbit clean walkthrough + full CI suite green confirm this.
4. **Test fixture safety**: Uses `os.tmpdir()` + `fs.mkdtemp` with proper `afterEach` cleanup. Matches the rest of `crud.test.ts`.

### Assessment

**Ready to merge: Yes, with one acknowledgment.**

**Reasoning:** Implementation correctly matches the Phase 3 spec; reader/writer separation is preserved; the helper is pure and isomorphic-safe; tests cover the documented branches and edge cases with clean fixtures. The reordering trade-off (Important #1) is inherent to the "keep last" rule — the test pins it down as expected behavior. Before merging, just confirm with the author that the divergent-content-with-same-uuid case (rare Syncthing scenario) is acceptable to silently collapse to the last record, with Phase 2's banner being the intended user-facing visibility for that. Minor items are post-merge cleanup if desired.

File: packages/core/src/session/crud.ts [hunk 1]
```diff
@@ -124,11 +124,25 @@ export const listSessions = (projectName: string) =>
     return sortSessionsByDate(sessions.filter((s): s is NonNullable<typeof s> => s !== null))
   })
 
-// Read session messages
+// Deduplicate messages by `uuid`, keeping the last occurrence. Messages without
+// a `uuid` (summary, file-history-snapshot, custom-title, agent-name) pass through
+// unchanged. Resolves Issue #137 Phase 3 — JSONL files with duplicated uuid records
+// (Syncthing conflicts, repeated history appends, cross-session edits) inflate the
+// rendered DOM and token-usage metrics; we keep only the latest record per uuid.
+export const dedupeMessagesByUuid = <T extends { uuid?: string }>(messages: T[]): T[] => {
```
Comment (DrumRobot) [inline packages/core/src/session/crud.ts:132]: **🟡 Important — Reordering acknowledgment**

When duplicates are interleaved with their own children, dedup keeps the last occurrence in its original position — children can end up before their parent in the array (see test `crud.test.ts:665-700`). `validateChain` is unaffected (uuid-set membership only), but sequential consumers (`MessageList.svelte`, tree renderers) may render orphaned children at the top.

For Syncthing conflicts with *divergent content* at the same uuid, the displayed conversation will show the assistant replying to a different message body than the one it actually replied to.

Recommend: add a JSDoc paragraph noting the trade-off. Phase 2 (UI warning banner) is the intended user-facing visibility for divergent-content cases.

File: packages/core/src/__tests__/crud.test.ts [hunk 2]
```diff
@@ -563,3 +563,181 @@ describe('updateMessageContent', () => {
...
+        message: { role: 'user', content: [{ type: 'text', text: 'dup-last' }] },
+      },
+    ])
+
+    const messages = await Effect.runPromise(readSession(projectName, sessionId))
+    expect(messages).toHaveLength(3)
+    expect(messages.map((m) => m.uuid)).toEqual(['user-1', 'assistant-1', 'dup'])
+    const dupMsg = messages.find((m) => m.uuid === 'dup')!
+    const payload = dupMsg.message as { content: Array<{ text: string }> }
+    expect(payload.content[0].text).toBe('dup-last')
```
Comment (DrumRobot) [inline packages/core/src/__tests__/crud.test.ts:700]: **🟡 Important — Missing chain-validity assertion**

The interleaved-duplicate test verifies the new order but does not assert chain validity. Adding one line documents that reordering is explicitly tolerated and would catch a future regression if `validateChain` ever started enforcing sequence:

```ts
const chainResult = validateChain(messages)
expect(chainResult.valid).toBe(true)
```
